File: exercises/java/src/main/java/com/adaptionsoft/games/uglytrivia/Game.java
package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Game {
    ArrayList players = new ArrayList();
    int[] places = new int[6];
    int[] coinCount = new int[6];
    boolean[] isInPenaltyBox = new boolean[6];

    LinkedList popQuestions = new LinkedList();
    LinkedList scienceQuestions = new LinkedList();
    LinkedList sportsQuestions = new LinkedList();
    LinkedList rockQuestions = new LinkedList();
    List<String> board = new ArrayList<>();
    static final int NUMBER_OF_QUESTIONS = 50;
    int currentPlayer = 0;
    boolean isGettingOutOfPenaltyBox;

    public Game() {
        for (int i = 0; i < NUMBER_OF_QUESTIONS; i++) {
            popQuestions.addLast(createQuestion(i,"Pop"));
            scienceQuestions.addLast(createQuestion(i,"Science"));
            sportsQuestions.addLast(createQuestion(i,"Sports"));
            rockQuestions.addLast(createQuestion(i,"Rock"));
        }
        for (int i = 0; i < 3; i++) {
            board.add("Pop");
            board.add("Science");
            board.add("Sports");
            board.add("Rock");
        }
    }

    public String createQuestion(int index, String category) {
        return ( category + " Question " + index);
    }

    public void addNewPlayer(String playerName) {
        players.add(playerName);
        places[numberOfPlayers()] = 0;
        coinCount[numberOfPlayers()] = 0;
        isInPenaltyBox[numberOfPlayers()] = false;

        System.out.println(playerName + " was added");
        System.out.println("They are player number " + players.size());

    }

    public int numberOfPlayers() {
        return players.size();
    }

    public void turn(int roll) {
        System.out.println(players.get(currentPlayer) + " is the current player");
        System.out.println("They have rolled a " + roll);
        if (isInPenaltyBox[currentPlayer]) {
            if (roll % 2 != 0) {
                isGettingOutOfPenaltyBox = true;
                System.out.println(players.get(currentPlayer) + " is getting out of the penalty box");
                movePlayer(roll);
                askQuestion();
            } else {
                System.out.println(players.get(currentPlayer) + " is not getting out of the penalty box");
                isGettingOutOfPenaltyBox = false;
            }

        } else {
            movePlayer(roll);
            askQuestion();
        }

    }

    private void askQuestion() {
        if (currentCategory() == "Pop")
            System.out.println(popQuestions.removeFirst());
        if (currentCategory() == "Science")
            System.out.println(scienceQuestions.removeFirst());
        if (currentCategory() == "Sports")
            System.out.println(sportsQuestions.removeFirst());
        if (currentCategory() == "Rock")
            System.out.println(rockQuestions.removeFirst());
    }

    private void movePlayer(int roll) {
        places[currentPlayer] = places[currentPlayer] + roll;
        if (places[currentPlayer] > board.size() - 1) places[currentPlayer] = places[currentPlayer] - board.size();
        System.out.println(players.get(currentPlayer)
                + "'s new location is "
                + places[currentPlayer]);
        System.out.println("The category is " + currentCategory());
    }

    private String currentCategory() {
        return board.get( places[currentPlayer]);
    }

    public boolean wasCorrectlyAnswered() {
        if (isInPenaltyBox[currentPlayer]) {
            if (isGettingOutOfPenaltyBox) {
                System.out.println("Answer was correct!!!!");
                return correctAnswer();
            } else {
                currentPlayer++;
                if (currentPlayer == players.size()) currentPlayer = 0;
                return true;
            }


        } else {

            System.out.println("Answer was corrent!!!!");
            return correctAnswer();
        }
    }

    private boolean correctAnswer() {
        coinCount[currentPlayer]++;
        System.out.println(players.get(currentPlayer)
                + " now has "
                + coinCount[currentPlayer]
                + " Gold Coins.");

        boolean winner = didPlayerWin();
        currentPlayer++;
        if (currentPlayer == players.size()) currentPlayer = 0;

        return winner;
    }

    public boolean wrongAnswer() {
        System.out.println("Question was incorrectly answered");
        System.out.println(players.get(currentPlayer) + " was sent to the penalty box");
        isInPenaltyBox[currentPlayer] = true;

        currentPlayer++;
        if (currentPlayer == players.size()) currentPlayer = 0;
        return true;
    }


    private boolean didPlayerWin() {
        return !(coinCount[currentPlayer] == 6);
    }
}

File: exercises/kotlin/src/main/kotlin/com/adaptionsoft/games/uglytrivia/Game.kt
package com.adaptionsoft.games.uglytrivia

import java.util.*

class Board {
    private val places = ArrayList<Place>()
    private val boardSize = 12

    init {
        createBoard(boardSize)
    }

    private fun createBoard(boardSize: Int) {
        for (i in 0..boardSize) {
            places.add(Place(i))
        }
    }

    fun getPlace(position: Int): Place {
        return places[position]
    }
}

class Player (val name: String){
    var purse: Int = 0
    var isInPenaltyBox: Boolean = false
    var position: Int = 0

    fun addCoin() {
        purse++
        println(
            "$name now has $purse Gold Coins."
        )
    }
}

class Place (position: Int) {
    private val questionCategory: Category

    init {
        questionCategory = when (position % 4) {
            0 -> Category.Pop
            1 -> Category.Science
            2 -> Category.Sports
            else -> Category.Rock
        }
    }
    fun category(): String {
        return questionCategory.toString()
    }
}

class QuestionBank {
    private val MAX_NUMBER_OF_QUESTIONS = 50

    internal var pop = LinkedList<Any>()
    internal var science = LinkedList<Any>()
    internal var sports = LinkedList<Any>()
    internal var rock = LinkedList<Any>()

    init {
        for (questionIndex in 0..<MAX_NUMBER_OF_QUESTIONS) {
            pop.addLast(createQuestion("Pop", questionIndex))
            science.addLast(createQuestion("Science", questionIndex))
            sports.addLast(createQuestion("Sports", questionIndex))
            rock.addLast(createQuestion("Rock", questionIndex))
        }
    }

    private fun createQuestion(category: String, index: Int): String {
        return "$category Question $index"
    }

    fun askQuestion(place: Place) {
        if (place.category() === "Pop")
            println(pop.removeFirst())
        if (place.category() === "Science")
            println(science.removeFirst())
        if (place.category() === "Sports")
            println(sports.removeFirst())
        if (place.category() === "Rock")
            println(rock.removeFirst())
    }
}

enum class Category {
    Pop, Science, Sports, Rock
}

class Game {
    private val NUMBER_OF_PLACES = 12
    private val COINS_TO_WIN = 6

    private val board = Board()
    private val questionBank = QuestionBank()
    private var players = ArrayList<Player>()

    private var currentPlayerIndex = 0
    private lateinit var currentPlayer: Player
    private var isGettingOutOfPenaltyBox: Boolean = false

    fun addPlayer(playerName: String): Boolean {
        players.add(Player(playerName))
        println("$playerName was added")
        println("They are player number " + players.size)
        currentPlayer = players[currentPlayerIndex]
        return true
    }

    fun roll(roll: Int) {
        println(currentPlayer.name + " is the current player")
        println("They have rolled a $roll")

        if (currentPlayer.isInPenaltyBox) {
            if (isRollOdd(roll)) {
                isGettingOutOfPenaltyBox = true
                println(currentPlayer.name + " is getting out of the penalty box")
                movePlayer(roll)
                questionBank.askQuestion(getCurrentPlace())
            } else {
                println(currentPlayer.name + " is not getting out of the penalty box")
                isGettingOutOfPenaltyBox = false
            }
        } else {
            movePlayer(roll)
            questionBank.askQuestion(getCurrentPlace())
        }
    }

    private fun isRollOdd(roll: Int) = roll % 2 != 0

    private fun movePlayer(roll: Int) {
        currentPlayer.position += roll
        if (currentPlayer.position >= NUMBER_OF_PLACES) currentPlayer.position -= NUMBER_OF_PLACES

        println(
            currentPlayer.name
                    + "'s new location is "
                    + currentPlayer.position
        )
        println("The category is " + getCurrentPlace().category())
    }

    private fun getCurrentPlace() = board.getPlace(currentPlayer.position)

    fun wasCorrectlyAnswered(): Boolean {
        if (currentPlayer.isInPenaltyBox) {
            if (isGettingOutOfPenaltyBox) {
                println("Answer was correct!!!!")
                currentPlayer.addCoin()
                val shouldContinueGame = shouldContinueGame()
                updateCurrentPlayer()
                return shouldContinueGame
            } else {
                updateCurrentPlayer()
                return true
            }
        } else {
            println("Answer was corrent!!!!")
            currentPlayer.addCoin()
            val shouldContinueGame = shouldContinueGame()
            updateCurrentPlayer()
            return shouldContinueGame
        }
    }

    fun wasIncorrectlyAnswered(): Boolean {
        println("Question was incorrectly answered")
        println(currentPlayer.name + " was sent to the penalty box")
        currentPlayer.isInPenaltyBox = true
        updateCurrentPlayer()
        return true
    }

    private fun updateCurrentPlayer() {
        currentPlayerIndex++
        if (currentPlayerIndex == players.size) currentPlayerIndex = 0
        currentPlayer = players[currentPlayerIndex]
    }

    private fun shouldContinueGame(): Boolean {
        return currentPlayer.purse != COINS_TO_WIN
    }
}